### PR TITLE
Relax matching criteria and patch question-related issues

### LINKED ISSUE #91: Implement getQuestionByPreference and update question services
Main task:
- [x] Implement getQuestionByPreference in FE question wrapper layer.
- [x] Consume new API from question service.

Others:
- [x] Fix constraints input field issues.
- [x] Update wrapper layer implementation.
- [x] Add UI feedback for event status.

Hold:
- [ ] Protect question routes. (wait for session authentication)

### PR DESCRIPTION
**Pull request by c0j0s**
# Pull Request

## Description
The matching logic now supports matching by common preference in `complexity` and `topic`. 
Upon matched, the success view should display their common preferences that will be used to query for questions.
After the owner requests to start the collab, the lobby will query for questions, and  redirect clients to collab page once the question is ready,

## Related Issue(s)
Close #91 

## Changes Made
- Relax matching criteria
- Support question selection for collab setup
- Add toast messages for the question page
- Update question wrapper layer
- Consume `getTopics` in matching card and question modal.
- Fix issue with hidden buttons in some matching lobby view
- Fix issue with the constraints and examples field in `ModifyQuestionModal`

## Screenshots (if applicable)

Matched view:
image

Question selection:
image

## Checklist
- [x] I have checked that the changes included in the PR are intended to merge to `master` or any destination branch.
- [x] I have verified that the new changes do not break any existing functionalities, unless the new changes are intended and have approved by the team.
- [x] I will take care of the merging and delete the side-branch after the PR is merged.

## Additional Notes/References
- Question wrapper did not implement `throwAndLogError` as server action are unable to throw custom exceptions to client component.

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -10,6 +10,12 @@
     "dev:collaboration-service": "cd ../backend/collaboration-service && npm run dev",
     "dev:matching-service": "cd ../backend/matching-service && npm run dev",
     "dev-all": "concurrently \"npm run dev:frontend\" \"npm run dev:user-service\" \"npm run dev:question-service\" \"npm run dev:matching-service\" \"npm run dev:collaboration-service\"",
+    "setup:frontend": "npm i",
+    "setup:user-service": "cd ../backend/user-service && npm i",
+    "setup:question-service": "cd ../backend/question-service && npm i",
+    "setup:collaboration-service": "cd ../backend/collaboration-service && npm i",
+    "setup:matching-service": "cd ../backend/matching-service && npm i",
+    "setup-all": "concurrently \"npm run setup:frontend\" \"npm run setup:user-service\" \"npm run setup:question-service\" \"npm run setup:matching-service\" \"npm run setup:collaboration-service\"",
```
Comment (tryyang2001) [inline frontend/package.json:18]: Thanks for creating this tool!!! Was about to create a PR to do that xD
@tlyi @xingjie99 @weidak for your reference.

Comment (weidak) [reply frontend/package.json:18]: 👍🏻

**Comment by tryyang2001** [comment]
Just an update of the bug that I reported yesterday. This is regarding the matching modal does not appear. Here is how to reproduce the bug:
1. For two users, try to match in the first round. Any topic is fine (I believe). For my case, I used (Python, Easy, Array). 
2. Once both parties accept the match and has been redirected to the collaboration workspace, exit the room by either clicking the end session button or click the logo to back to home page. 
3. Try clicking the `Get Matched` button again, and the modal will not appear.

Possible guess: 
I think it might be related to the need of doing page refresh. Because every time when I encountered this, refresh the page and it will back to normal.

**Comment by tryyang2001** [comment]
Another flaw that I found is that even though the two parties have clicked the ready button, if one person refreshed his page, then he gets disconnected. Not sure if this is the intended behaviour? One issue might be that the room host click the start peerprep button but then like 0.001s before the button click the user refreshes his page... ?
image
After `John Doe` refresh his page:
image

Ah and another potential issue from the user experience POV is that if both parties have clicked the ready button, but the room host does not click the `start peerprep` button and then the partner can only wait haha. Ok even though for now he can simply quit the session by page refreshing. So if this page refresh is not intended and then after you fixed that issue, you might also want to consider setting timeout/auto-start for this issue.

**Comment by tryyang2001** [comment]
Ui issue again:
As for now if the topic tags are too many, then the overflowed topic cannot be seen. Do you think if it is possible to allow horizontal scrolling in your dropdown list?
image

**Comment by tryyang2001** [comment]
Another bug that I found is that when this thing comes out:
image
Entering the collaboration page from the room host side will have this error, but this error does not break any functionality:
image

If you want to reproduce this error:
1. For two users A and B, set A preferences as (Python, Medium, Array, String), and B preferences as (Python, Medium, String). Let's us assume that A is the room host.
2. Accept the match, pick any of the question in the dropdown list, and enter the collaboration workspace.
3. You will see the error toast occurs at A, this only occurs at the host side.

So I have a follow-up question regarding this issue:
I was actually thinking that can we do random question assignment instead of letting the room host pick a question? So that when we have the history service, we can remove questions that are already attempted by either one of the room participants, or at least reduce the likelihood of getting attempted questions.

**Comment by c0j0s** [comment]
Thankss @tryyang2001 for finding the bugs.

> Just an update of the bug that I reported yesterday. This is regarding the matching modal does not appear. Here is how to reproduce the bug:

Fixed

> Another flaw that I found is that even though the two parties have clicked the ready button, if one person refreshed his page, then he gets disconnected. Not sure if this is the intended behaviour? One issue might be that the room host click the start peerprep button but then like 0.001s before the button click the user refreshes his page... ?

It is the intended behavior. Once the page got refreshed, the socket gets disconnected, even if the user request matching again, the socket Id is different already, so there's no way to rejoin the user unless matching service have knowledge of the user database. so I guess the simplest approach is to disconnect them.

> Ui issue again:

added a tooptip for it.

> I was actually thinking that can we do random question assignment instead of letting the room host pick a question? So that when we have the history service, we can remove questions that are already attempted by either one of the room participants, or at least reduce the likelihood of getting attempted questions.

no idea what cause the runtime error. as for random assignment I dont think it matters, I can filter the options when history service is ready. but that raises another question is that who's history should we take into account.

> server action are unable to throw custom exceptions to client component.

let's continue this discussion over #47

**Review by tryyang2001 — APPROVED** [review]
LGTM, the matching now looks very cool!!

**Pull request merged**